dataStructures/queues/queue.go

```diff
@@ -45,25 +45,42 @@ func (q *Queue) Add(order *Order) {
 		*q = Queue{order}
 	} else {
 		var appended bool
+		var insert int
+
+		// using binary-search to know index to insert the order.
+		lowIndex := 0
+		highIndex := len(*q) - 1
 
 	QLoop:
-		for i, existingOrder := range *q {
-			var insert int
+		for lowIndex <= highIndex {
+
+			midPoint := (lowIndex + highIndex) / 2
+			guessOrder := (*q)[midPoint]
 
 			switch {
-			case order.priority < existingOrder.priority:
-				insert = i
-			case order.priority == existingOrder.priority:
-				insert = 1 + i
-				// peek to check next order in the queue.
-				if insert < len(*q) && (*q)[insert].priority == order.priority {
-					// keep looking for the best location to insert.
-					continue QLoop
-				}
+			case order.priority < guessOrder.priority:
+				// we move closer to the left-side
+				highIndex = midPoint - 1
+			case order.priority > guessOrder.priority:
+				// we move to the right-side of the queue
+				lowIndex = midPoint + 1
```

Comment:
This looks like where your bug could be. The reason that binary search is `O(log n)` is because you split it in half each time and that takes `log_2(n)` splits (inverse of `2^s=n` where `s` is the number of splits). What you're doing here is only shrinking by one, but your checks for `<` or `>` can rule out everything on one side of the guess, so you can set `{high,low}Index` to `midPoint`!

Reply:
Thanks for spotting this, I will work on it.